### PR TITLE
TCTI: Add USB TPM (LetsTrust-TPM2Go) TCTI module

### PR DESCRIPTION
**Pull request by wxleong**
Bringing up a TCTI module to support the LetsTrust-TPM2Go, a USB TPM from @PaulKissinger. It's a combined effort and resources are taken from:
- @PeterHuewe: https://github.com/PeterHuewe/tpm2-tss/tree/letstrust2go, https://github.com/PeterHuewe/tpm2-tss/tree/letstrust2go-wo-waitstates
- @tobyp: https://gist.github.com/tobyp/aed5598188088f4abbeb737b408e5287

The CI is expected to fail at this stage since the library `libusb-1.0-0-dev` is missing from the CI container image.

### CODE REVIEW

**Comment by williamcroberts** [comment]
> Apparently, the build is still failing on Fedora-32 and Alpine-3.15.
> 
> This is weird, because of https://github.com/tpm2-software/tpm2-software-container/blob/f15f2ee3e4d69ac332c8acfaa8867d0546090cda/fedora-32.docker.m4#L62 and https://github.com/tpm2-software/tpm2-software-container/blob/f15f2ee3e4d69ac332c8acfaa8867d0546090cda/alpine-3.15.docker.m4#L55
> 
> I thought, the dockers were updated automatically?
> 
> @williamcroberts Do you have any idea ?

Their's a rule in the build do only build/and publish for containers detected in the diff, but the diff action seems to have a bug and only compares the top level commit (the build output said this is the valid compare url):
  - https://github.com/tpm2-software/tpm2-software-container/compare/c081d3ea5d26...f15f2ee3e4d6
In that diff, their is no changes to published containers so the publish is all skipped.

**Comment by wxleong** [comment]
@williamcroberts CI workflow has passed, but CIFuzz and CirrusCI workflows still encountered libusb-1.0 library missing error.

**Comment by williamcroberts** [comment]
> @williamcroberts CI workflow has passed, but CIFuzz and CirrusCI workflows still encountered libusb-1.0 library missing error.

@wxleong for
CI Fuzz, add it to the dockerfile here:
- https://github.com/google/oss-fuzz/tree/aadb0925eef67d068ef2b66fa44764496e4d4493/projects/tpm2-tss

For CirrusCI add it to this file (Note it's FreeBSD):
- https://github.com/tpm2-software/tpm2-tss/blob/master/.cirrus.yml

**Comment by wxleong** [comment]
Last item to fix: CIFuzz, simple fix but have to sort out the Google Contributor License Agreement before I can do anything.

**Comment by PaulKissinger** [comment]
Hi WenXin, 

Thanks formall this efford!

Happy holidays if you have one 😅

Paul

-----Original Message-----

**Comment by williamcroberts** [comment]
> Last item to fix: CIFuzz, simple fix but have to sort out the Google Contributor License Agreement before I can do anything.

I can do CI Fuzz

**Comment by williamcroberts** [comment]
Lets take this now, as their will be a chicken and egg situation with google's OSS fuzz.

**Comment by williamcroberts** [comment]
@wxleong:
1. I sent the PR to oss-fuzz here: https://github.com/google/oss-fuzz/pull/9264
2. Can you rebase this on-top of current master and fix the conflicts and re-push
    a. `git pull --rebase origin master`
    b. fix merge conflicts
    c. commit
    d. `git push -f wxleong develop-tcti-spi-lt2go3; # You may need to change the names to match what you have`

Thanks,
Bill

**Comment by wxleong** [comment]
@williamcroberts all looks good now

**Comment by wxleong** [comment]
> @wxleong:
> 
> 1. I sent the PR to oss-fuzz here: [tpm2-tss: disable spi lt2go TCTI google/oss-fuzz#9264](https://github.com/google/oss-fuzz/pull/9264)
> 2. Can you rebase this on-top of current master and fix the conflicts and re-push
>    a. `git pull --rebase origin master`
>    b. fix merge conflicts
>    c. commit
>    d. `git push -f wxleong develop-tcti-spi-lt2go3; # You may need to change the names to match what you have`
> 
> Thanks, Bill

@williamcroberts thanks for the help with oss-fuzz. The conflicts was already resolved prior to this. I did a dummy force push and all are passing now.

**Comment by tobyp** [comment]
Awesome! I look forward to giving this a spin after Christmas.

File: src/tss2-tcti/tcti-spi-lt2go.c
Comment (williamcroberts) [inline src/tss2-tcti/tcti-spi-lt2go.c:186]: here and everywhere else `/* */ ` for comments.

File: src/tss2-tcti/tcti-spi-lt2go.c
Comment (williamcroberts) [inline src/tss2-tcti/tcti-spi-lt2go.c:196]: no newline here, use calloc and drop memset below.

File: src/tss2-tcti/tcti-spi-lt2go.c
Comment (williamcroberts) [inline src/tss2-tcti/tcti-spi-lt2go.c:201]: use calloc and drop this memset

**Review by williamcroberts — CHANGES_REQUESTED** [review]
I think you also want to rebase ontop of current master

File: .cirrus.yml
Comment (wxleong) [reply .cirrus.yml:27]: Two reason behind this
1. Not available via pkg install directly
2. FreeBSD has libusb preinstalled, however it is an older incompatible version LIBUSB_API_VERSION (0x01000102), the tcti module requires 0x01000109

File: configure.ac
Comment (wxleong) [reply configure.ac]: done

File: src/tss2-tcti/tcti-spi-lt2go.c
Comment (wxleong) [reply src/tss2-tcti/tcti-spi-lt2go.c]: fixed

File: src/tss2-tcti/tcti-spi-lt2go.c
Comment (wxleong) [reply src/tss2-tcti/tcti-spi-lt2go.c:198]: fixed

File: src/tss2-tcti/tcti-spi-lt2go.c
Comment (wxleong) [reply src/tss2-tcti/tcti-spi-lt2go.c]: fixed

**Comment by wxleong** [comment]
> I think you also want to rebase ontop of current master

merged to master 2b7007c

**Comment by williamcroberts** [comment]
> > I think you also want to rebase ontop of current master
> 
> merged master, [2b7007c](https://github.com/tpm2-software/tpm2-tss/commit/2b7007ce37b9063144c169adcf0dbb8219b16c33)

Again, always rebase, if your using git directly its:
```bash
git pull --rebase origin master
```
Ill update this one as well, same instructions for updating as:
  - https://github.com/tpm2-software/tpm2-tss/pull/2503

Also, ill squash the history down to one comimt, we don't want commits that are like:
1. add feature X
2. fix thing A
3. fix thing B

It should all be atomic.

File: Makefile-test.am
Comment (williamcroberts) [inline Makefile-test.am:538]: This block needs to be in the AM conditional block as well:
```
if ENABLE_TCTI_SPI_LT2GO
...
endif
```

**Review by williamcroberts — CHANGES_REQUESTED** [review]
Please force reset you branch and amend the history, as you make a change related to the PR git commit --amend. If you need to update the branch ontop of master, git pull --rebase origin master is your friend as well.

**Comment by tobyp** [comment]
I was able to build this on ArchLinux, but it fails for me with "Probing TPM failed" because spi_tpm_helper_read_reg(TCTI_SPI_HELPER_TPM_DID_VID_REG) keeps reading 0. Logs with tcti+TRACE are [here](https://gist.github.com/tobyp/9ea2a3a2a553e5d5a3186c114f1b4223) if you'd like to have a look.

@PaulKissinger have there been any changes to the hardware? I have the relatively old stick with serial number 000008 here.

**Comment by PaulKissinger** [comment]
> I was able to build this on ArchLinux, but it fails for me with "Probing TPM failed" because spi_tpm_helper_read_reg(TCTI_SPI_HELPER_TPM_DID_VID_REG) keeps reading 0. Logs with tcti+TRACE are [here](https://gist.github.com/tobyp/9ea2a3a2a553e5d5a3186c114f1b4223) if you'd like to have a look.
> 
> @PaulKissinger have there been any changes to the hardware? I have the relatively old stick with serial number 000008 here.

Hi  @tobyp,
It is possible that you have the wrong hardware.
May you send me a detailed picture of the area between the chips with parts of both chips?

Thanks for testing!

Paul

**Comment by williamcroberts** [comment]
> I also see an error message in the kernel log for every tss command, like this:
> 
> ```
> [ 9695.096293] usb 1-1.3: usbfs: process 108052 (tpm2_startup) did not claim interface 0 before use
> ```
> 
> I made a fix commit [here](https://github.com/tobyp/tpm2-tss/commit/ea21839e64b3863e4b5ed21daa81c7da970ae206), feel free to squash it into your branch if you agree this is a good solution (I'm not really fit with USB, I'm not even sure what "claiming" an interface means, just that the error bothers me :P).

I'm not fit either, but I don't see anything glairingly wrong... I'll let @wxleong weigh in and we can pick it in.

**Comment by tobyp** [comment]
> Hi @tobyp, It is possible that you have the wrong hardware. May you send me a detailed picture of the area between the chips with parts of both chips?

letstrust_tpm2go

I hope this is detailed enough, else I'll have to find a better camera than my phone :sweat_smile:

**Comment by PaulKissinger** [comment]
Hi @tobyp, 

yes, wrong hardware revision. The chip select of the TPM is connected to the GPIO1 of the USB Bridge. I have a new revision on my desk so I'll replace your stick in the next weeks.

Context: The patchset from @wxleong uses the burst count register and so wait states are not needed. This behavior reduces the communication between the host and the USB bridge. In the early phases of the hardware, it was not clear if the Chipselect works as expected, or that these sticks need a user-controllable chip select via GPIO. The latest HW revision will have a USB Bridge-controlled chip select. 

Thank you and best regards,
 
Paul

File: Makefile-test.am
Comment (wxleong) [reply Makefile-test.am]: fixed

File: Makefile-test.am
Comment (wxleong) [reply Makefile-test.am:541]: fixed

**Comment by wxleong** [comment]
> I also see an error message in the kernel log for every tss command, like this:
> 
> ```
> [ 9695.096293] usb 1-1.3: usbfs: process 108052 (tpm2_startup) did not claim interface 0 before use
> ```
> 
> I made a fix commit [here](https://github.com/tobyp/tpm2-tss/commit/ea21839e64b3863e4b5ed21daa81c7da970ae206), feel free to squash it into your branch if you agree this is a good solution (I'm not really fit with USB, I'm not even sure what "claiming" an interface means, just that the error bothers me :P).

I'll look into this

**Comment by wxleong** [comment]
> > > I think you also want to rebase ontop of current master
> > 
> > 
> > merged master, [2b7007c](https://github.com/tpm2-software/tpm2-tss/commit/2b7007ce37b9063144c169adcf0dbb8219b16c33)
> 
> Again, always rebase, if your using git directly its:
> 
> ```shell
> git pull --rebase origin master
> ```
> 
> Ill update this one as well, same instructions for updating as:
> 
> * [TCTI: Add USB TPM (FTDI MPSSE USB to SPI bridge) TCTI module #2503](https://github.com/tpm2-software/tpm2-tss/pull/2503)
> 
> Also, ill squash the history down to one comimt, we don't want commits that are like:
> 
> 1. add feature X
> 2. fix thing A
> 3. fix thing B
> 
> It should all be atomic.

Thanks for the help on squashing the commits. Merge was a mistake, moving forward i'll stick to rebase.

**Comment by wxleong** [comment]
Looks like another "libusb-1.0 library is missing" error with the newly added CodeQL

**Comment by wxleong** [comment]
> > I also see an error message in the kernel log for every tss command, like this:
> > ```
> > [ 9695.096293] usb 1-1.3: usbfs: process 108052 (tpm2_startup) did not claim interface 0 before use
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I made a fix commit [here](https://github.com/tobyp/tpm2-tss/commit/ea21839e64b3863e4b5ed21daa81c7da970ae206), feel free to squash it into your branch if you agree this is a good solution (I'm not really fit with USB, I'm not even sure what "claiming" an interface means, just that the error bothers me :P).
> 
> I'll look into this

Fixed in [9b97f20](https://github.com/tpm2-software/tpm2-tss/commit/9b97f205eee02710709ce402a59611d82e8c5b5a) thanks @tobyp.

**Comment by wxleong** [comment]
@williamcroberts all requested changes are applied, please review

**Review by williamcroberts — APPROVED** [review]
I am good with this, I'll let @JuergenReppSIT do a review and merge since I am stepping down as a full time maintainer.

**Pull request merged**
